pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -114,64 +120,69 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
 			// build listener with tcp proxy, with or without TLS context, for TCP servers
 			//   or TLS servers using simple/mutual/passthrough TLS
 			//   or HTTPS servers using passthrough TLS
 			// This process typically yields multiple filter chain matches (with SNI) [if TLS is used]
-			filterChainOpts := make([]*filterChainOpts, 0)
+			newFilterChainOpts := make([]*filterChainOpts, 0)
 
 			for _, server := range servers {
 				if gateway.IsTLSServer(server) && gateway.IsHTTPServer(server) {
 					routeName := mergedGateway.TLSServerInfo[server].RouteName
 					// This is a HTTPS server, where we are doing TLS termination. Build a http connection manager with TLS context
-					filterChainOpts = append(filterChainOpts, configgen.createGatewayHTTPFilterChainOpts(builder.node, server.Port, server,
+					newFilterChainOpts = append(newFilterChainOpts, configgen.createGatewayHTTPFilterChainOpts(builder.node, server.Port, server,
 						routeName, proxyConfig))
-					filterChains = append(filterChains, istionetworking.FilterChain{
+					newFilterChains = append(newFilterChains, istionetworking.FilterChain{
 						ListenerProtocol:   istionetworking.ListenerProtocolHTTP,
 						IstioMutualGateway: server.Tls.Mode == networking.ServerTLSSettings_ISTIO_MUTUAL,
 					})
 				} else {
-					// passthrough or tcp, yields multiple filter chains
+					// This is the case of TCP or PASSTHROUGH.
 					tcpChainOpts := configgen.createGatewayTCPFilterChainOpts(builder.node, builder.push,
 						server, mergedGateway.GatewayNameForServer[server])
-					filterChainOpts = append(filterChainOpts, tcpChainOpts...)
+					newFilterChainOpts = append(newFilterChainOpts, tcpChainOpts...)
 					for i := 0; i < len(tcpChainOpts); i++ {
-						filterChains = append(filterChains, istionetworking.FilterChain{ListenerProtocol: istionetworking.ListenerProtocolTCP})
+						newFilterChains = append(newFilterChains, istionetworking.FilterChain{ListenerProtocol: istionetworking.ListenerProtocolTCP})
 					}
 				}
 			}
-			opts.filterChainOpts = filterChainOpts
+			opts.filterChainOpts = newFilterChainOpts
 		}
 
-		l := buildListener(opts, core.TrafficDirection_OUTBOUND)
-
-		mutable := &MutableListener{
-			MutableObjects: istionetworking.MutableObjects{
-				Listener: l,
-				// Note: buildListener creates filter chains but does not populate the filters in the chain; that's what
-				// this is for.
-				FilterChains: filterChains,
-			},
+		var mutable *MutableListener
+		if mopts, exists := mutableopts[lname]; !exists {
+			mutable = &MutableListener{
+				MutableObjects: istionetworking.MutableObjects{
+					// Note: buildListener creates filter chains but does not populate the filters in the chain; that's what
+					// this is for.
+					FilterChains: newFilterChains,
+				},
+			}
+			mutableopts[lname] = mutableListenerOpts{mutable: mutable, opts: opts}
+		} else {
+			mopts.opts.filterChainOpts = append(mopts.opts.filterChainOpts, opts.filterChainOpts...)
```

Comment:
There is a small bug here. It definitely exists in current master, not sure if it did in 1.9. The iteration of MergedServers is randomized order due to golang maps, so we may get different filter chains. This doesn't matter to envoy logic, but *might* trigger incorrect behavior in some places like envoy diffing, istioctl diffing, istiod caching (not a concern now or in the near future, but in theory...)